src/zenml/config/compiler.py

```diff
@@ -648,6 +653,103 @@ def _compute_pipeline_spec(
             "parameters": pipeline._parameters,
         }
 
+        # Store a JSON schema of the pipeline input parameters for CLI/UI
+        # tooling.
+        if pipeline_parameters_model := pipeline.get_parameters_model():
+            additional_spec_args["parameters_schema"] = (
+                pipeline_parameters_model.model_json_schema()
+            )
+
+        # Best-effort: build a response schema snapshot for step outputs (tooling).
+        try:
+            # Map invocation id -> StepSpec for quick lookup
+            name_to_spec: Dict[str, StepSpec] = {
+                s.pipeline_parameter_name: s for s in step_specs
+            }
+            all_names: set[str] = set(name_to_spec.keys())
+            # Build downstream map from upstream relationships
+            downstream: Dict[str, Set[str]] = {n: set() for n in all_names}
+            for s in step_specs:
+                for up in s.upstream_steps:
+                    if up in downstream:
+                        downstream[up].add(s.pipeline_parameter_name)
+            # NOTE: The serving response uses flat keys in the form
+            # "{step}.{output}". We therefore build a flat outputs schema
+            # instead of a nested per-step structure.
+            outputs_properties: Dict[str, Any] = {}
+            all_defs: Dict[str, Any] = {}
+
+            # Include all steps to reflect actual serving outputs
+            for name in all_names:
+                spec = name_to_spec[name]
+                step_instance = BaseStep.load_from_source(spec.source)
+                out_sigs = parse_return_type_annotations(
+                    func=step_instance.entrypoint
+                )
+                if not out_sigs:
+                    continue
+                for out_name, sig in out_sigs.items():
+                    try:
+                        ta = TypeAdapter(sig.resolved_annotation)
+                        schema = ta.json_schema()
+                        if "$defs" in schema:
+                            all_defs.update(schema["$defs"])
+                            schema = {
+                                k: v for k, v in schema.items() if k != "$defs"
+                            }
+                        # Improve UI example for generic object schemas to avoid
+                        # the 'additionalProp1' placeholder in Swagger UI.
+                        if (
+                            isinstance(schema, dict)
+                            and schema.get("type") == "object"
+                            and "properties" not in schema
+                        ):
+                            schema.setdefault("example", {})
+                        # Flat key matches serving response shape
+                        outputs_properties[f"{name}.{out_name}"] = schema
```

Comment:
I'm wondering what the purpose of this code is. This is now parsing **all outputs** of **all steps** and includes them in the output spec? Why would the output of some intermediate step be part of the output spec of a pipeline? If anything, that should be part of each step spec separately?

Reply:
valid point updated this to only handle pipeline output instead